File: src/main/java/pl/yalgrin/playnite/simplesync/aop/ReactiveLoggingAspect.java
package pl.yalgrin.playnite.simplesync.aop;

import org.aspectj.lang.ProceedingJoinPoint;
import org.aspectj.lang.Signature;
import org.aspectj.lang.annotation.Around;
import org.aspectj.lang.annotation.Aspect;
import org.aspectj.lang.annotation.Pointcut;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Component;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;

import java.util.Arrays;
import java.util.List;
import java.util.Optional;
import java.util.UUID;
import java.util.concurrent.atomic.AtomicInteger;

import static pl.yalgrin.playnite.simplesync.security.ClientUtilKt.getSessionClientId;

@Aspect
@Component
public class ReactiveLoggingAspect {

    @Pointcut("execution(* pl.yalgrin.playnite.simplesync.service..*.*(..))")
    public void servicePointcut() {
    }

    @Pointcut("execution(* pl.yalgrin.playnite.simplesync.web..*.*(..))")
    public void resourcePointcut() {
    }

    @Around("servicePointcut() || resourcePointcut()")
    public Object logAround(ProceedingJoinPoint joinPoint) throws Throwable {
        Signature signature = joinPoint.getSignature();
        String name = signature.getName();
        Object[] args = joinPoint.getArgs();
        Object returnValue = joinPoint.proceed();
        Class<?> loggerClass = Optional.ofNullable(joinPoint.getTarget()).map(Object::getClass)
                .orElseGet(signature::getDeclaringType);
        Logger logger = LoggerFactory.getLogger(loggerClass);
        if (returnValue != null) {
            if (returnValue instanceof Mono<?> mono) {
                UUID uuid = UUID.randomUUID();
                return getSessionClientId()
                        .defaultIfEmpty("(none)")
                        .flatMap(clientId -> mono
                                .doOnSubscribe(
                                        _ -> logger.debug("{} > START, uuid: {}, clientId: {}, args: {}", name, uuid,
                                                clientId, formatArgs(args)))
                                .doOnSuccess(
                                        result -> logger.debug("{} > END, uuid: {}, clientId: {}, result: {}", name,
                                                uuid, clientId,
                                                formatResult(result)))
                                .doOnError(
                                        th -> logger.error("{} > ERROR, uuid: {}, clientId: {}", name, uuid, clientId,
                                                th))
                                .doOnCancel(() -> logger.debug("{} > CANCEL, uuid: {}, clientId: {}", name, uuid,
                                        clientId)));
            }
            if (returnValue instanceof Flux<?> flux) {
                UUID uuid = UUID.randomUUID();
                AtomicInteger counter = new AtomicInteger(0);
                return getSessionClientId()
                        .defaultIfEmpty("(none)")
                        .flatMapMany(clientId -> flux
                                .doOnSubscribe(
                                        _ -> logger.debug("{} > START, uuid: {}, clientId: {}, args: {}", name, uuid,
                                                clientId, formatArgs(args)))
                                .doOnNext(_ -> counter.incrementAndGet())
                                .doOnComplete(
                                        () -> logger.debug("{} > END, uuid: {}, clientId: {}, result.size(): {}", name,
                                                uuid,
                                                clientId, counter.get()))
                                .doOnError(
                                        th -> logger.error("{} > ERROR, uuid: {}, clientId: {}", name, uuid, clientId,
                                                th))
                                .doOnCancel(() -> logger.debug("{} > CANCEL, uuid: {}, clientId: {}", name, uuid,
                                        clientId)));
            }
        }
        return returnValue;
    }

    private static List<Object> formatArgs(Object[] args) {
        return Arrays.stream(args).map(ReactiveLoggingAspect::formatResult).toList();
    }

    private static Object formatResult(Object result) {
        if (result instanceof byte[] arr) {
            return "byte[" + arr.length + "]";
        }
        return result;
    }
}

File: src/main/kotlin/pl/yalgrin/playnite/simplesync/common/aop/ReactiveLoggingAspect.kt
package pl.yalgrin.playnite.simplesync.common.aop

import org.aspectj.lang.ProceedingJoinPoint
import org.aspectj.lang.Signature
import org.aspectj.lang.annotation.Around
import org.aspectj.lang.annotation.Aspect
import org.aspectj.lang.annotation.Pointcut
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Component
import pl.yalgrin.playnite.simplesync.security.getSessionClientId
import reactor.core.publisher.Flux
import reactor.core.publisher.Mono
import java.util.*
import java.util.concurrent.atomic.AtomicInteger

@Aspect
@Component
class ReactiveLoggingAspect {
    @Pointcut("execution(* pl.yalgrin.playnite.simplesync.service..*.*(..)) || execution(* pl.yalgrin.playnite.simplesync.*.service..*.*(..))")
    fun servicePointcut() = Unit

    @Pointcut("execution(* pl.yalgrin.playnite.simplesync.web..*.*(..))")
    fun resourcePointcut() = Unit

    @Around("servicePointcut() || resourcePointcut()")
    @Throws(Throwable::class)
    fun logAround(joinPoint: ProceedingJoinPoint): Any? {
        val signature: Signature = joinPoint.signature
        val name: String = signature.name
        val args: Array<Any?> = joinPoint.args
        val returnValue: Any? = joinPoint.proceed()
        val loggerClass: Class<*> = joinPoint.target?.javaClass ?: signature.declaringType
        val logger: Logger = LoggerFactory.getLogger(loggerClass)
        if (returnValue != null) {
            if (returnValue is Mono<*>) {
                val uuid: UUID = UUID.randomUUID()
                return getSessionClientId().defaultIfEmpty("(none)").flatMap { clientId ->
                    returnValue.doOnSubscribe {
                        logger.debug(
                            "{} > START, uuid: {}, clientId: {}, args: {}",
                            name,
                            uuid,
                            clientId,
                            formatArgs(args)
                        )
                    }.doOnSuccess { result ->
                        logger.debug(
                            "{} > END, uuid: {}, clientId: {}, result: {}",
                            name,
                            uuid,
                            clientId,
                            formatResult(result)
                        )
                    }.doOnError { th ->
                        logger.error(
                            "{} > ERROR, uuid: {}, clientId: {}", name, uuid, clientId, th
                        )
                    }.doOnCancel {
                        logger.debug("{} > CANCEL, uuid: {}, clientId: {}", name, uuid, clientId)
                    }
                }
            }
            if (returnValue is Flux<*>) {
                val uuid: UUID = UUID.randomUUID()
                val counter = AtomicInteger(0)
                return getSessionClientId().defaultIfEmpty("(none)").flatMapMany { clientId ->
                    returnValue.doOnSubscribe {
                        logger.debug(
                            "{} > START, uuid: {}, clientId: {}, args: {}",
                            name,
                            uuid,
                            clientId,
                            formatArgs(args)
                        )
                    }.doOnNext { counter.incrementAndGet() }.doOnComplete {
                        logger.debug(
                            "{} > END, uuid: {}, clientId: {}, result.size(): {}",
                            name,
                            uuid,
                            clientId,
                            counter.get()
                        )
                    }.doOnError { th ->
                        logger.error("{} > ERROR, uuid: {}, clientId: {}", name, uuid, clientId, th)
                    }.doOnCancel {
                        logger.debug("{} > CANCEL, uuid: {}, clientId: {}", name, uuid, clientId)
                    }
                }
            }
        }
        return returnValue
    }

    private fun formatArgs(args: Array<Any?>): List<Any?> {
        return args.map { formatResult(it) }
    }

    private fun formatResult(result: Any?): Any? {
        return if (result is ByteArray) {
            "byte[${result.size}]"
        } else {
            result
        }
    }
}
